File: apps/cptr/cptr/services/mcp_activity.py
"""Bounded in-memory MCP tool activity for the admin Console.

Unlike ``mcp_traffic``, this channel may carry already-redacted bounded tool
input/output prepared by the MCP adapter. It remains non-durable and accepts
only a strict allowlisted event envelope.
"""

from __future__ import annotations

import asyncio
import os
from collections import deque
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class McpActivityClient(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=128)
    label: str = Field(min_length=1, max_length=80)
    version: str | None = Field(default=None, max_length=64)


class McpActivityEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: Literal[1]
    event_id: str = Field(min_length=8, max_length=128)
    sequence: int = Field(ge=1)
    timestamp_ms: int = Field(ge=0)
    client: McpActivityClient
    session_id: str | None = Field(default=None, max_length=128)
    request_id: str | None = Field(default=None, max_length=128)
    correlation_id: str | None = Field(default=None, max_length=128)
    tool_name: str = Field(min_length=1, max_length=256)
    title: str | None = Field(default=None, max_length=160)
    phase: Literal["started", "complete", "failed"]
    summary: str = Field(min_length=1, max_length=500)
    arguments_json: str | None = Field(default=None, max_length=13_000)
    result_json: str | None = Field(default=None, max_length=13_000)
    error_json: str | None = Field(default=None, max_length=13_000)
    duration_ms: int | None = Field(default=None, ge=0, le=86_400_000)


class McpActivityBatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    events: list[McpActivityEvent] = Field(min_length=1, max_length=100)


def _bounded_env_int(name: str, default: int, minimum: int, maximum: int) -> int:
    raw = os.getenv(name)
    if raw is None:
        return default
    try:
        value = int(raw)
    except (TypeError, ValueError):
        return default
    return max(minimum, min(value, maximum))


class McpActivityStore:
    """Small bounded event store with non-blocking subscriber fan-out."""

    def __init__(self, *, max_events: int = 250, subscriber_queue_size: int = 64) -> None:
        self.max_events = max(1, int(max_events))
        self.subscriber_queue_size = max(1, int(subscriber_queue_size))
        self._dedupe_capacity = max(32, self.max_events * 4)
        self._lock = asyncio.Lock()
        self._events: deque[dict[str, object]] = deque(maxlen=self.max_events)
        self._seen_event_ids: deque[str] = deque()
        self._seen_event_id_set: set[str] = set()
        self._subscribers: set[asyncio.Queue[dict[str, object]]] = set()
        self._ingestion_sequence = 0
        self._slow_subscriber_drops = 0

    def subscribe(self) -> asyncio.Queue[dict[str, object]]:
        queue: asyncio.Queue[dict[str, object]] = asyncio.Queue(maxsize=self.subscriber_queue_size)
        self._subscribers.add(queue)
        return queue

    def unsubscribe(self, queue: asyncio.Queue[dict[str, object]]) -> None:
        self._subscribers.discard(queue)

    async def ingest(self, events: list[McpActivityEvent]) -> dict[str, int]:
        accepted = 0
        duplicates = 0
        dropped = 0
        async with self._lock:
            for event in events:
                if event.event_id in self._seen_event_id_set:
                    duplicates += 1
                    continue
                self._remember_event_id(event.event_id)
                self._ingestion_sequence += 1
                projected = event.model_dump(mode="json")
                projected["ingestion_sequence"] = self._ingestion_sequence
                self._events.append(projected)
                self._fan_out(projected)
                accepted += 1
        return {"accepted": accepted, "duplicates": duplicates, "dropped": dropped}

    async def snapshot(self) -> dict[str, object]:
        async with self._lock:
            return {
                "version": 1,
                "sequence": self._ingestion_sequence,
                "events": list(self._events),
                "stream_health": {
                    "subscriber_count": len(self._subscribers),
                    "slow_subscriber_drops": self._slow_subscriber_drops,
                    "event_capacity": self.max_events,
                    "subscriber_queue_capacity": self.subscriber_queue_size,
                },
            }

    def _remember_event_id(self, event_id: str) -> None:
        self._seen_event_ids.append(event_id)
        self._seen_event_id_set.add(event_id)
        while len(self._seen_event_ids) > self._dedupe_capacity:
            expired = self._seen_event_ids.popleft()
            self._seen_event_id_set.discard(expired)

    def _fan_out(self, projected: dict[str, object]) -> None:
        for queue in tuple(self._subscribers):
            try:
                queue.put_nowait(projected)
                continue
            except asyncio.QueueFull:
                pass

            try:
                queue.get_nowait()
                self._slow_subscriber_drops += 1
            except asyncio.QueueEmpty:
                pass
            try:
                queue.put_nowait(projected)
            except asyncio.QueueFull:
                self._slow_subscriber_drops += 1


mcp_activity_store = McpActivityStore(
    max_events=_bounded_env_int("CPTR_MCP_ACTIVITY_MAX_EVENTS", 250, 25, 2000),
    subscriber_queue_size=_bounded_env_int("CPTR_MCP_ACTIVITY_SUBSCRIBER_QUEUE", 64, 8, 512),
)

File: apps/cptr/cptr/frontend/src/lib/stores/mcp-activity.ts
import type {
	McpActivityEvent,
	McpActivityPhase,
	McpActivitySnapshot,
	McpContentItem
} from '$lib/apis/mcp';

export type McpActivitySource = 'plugin' | 'console';

export type McpActivityRow = {
	id: string;
	correlationKey: string;
	correlationId?: string | null;
	source: McpActivitySource;
	sequence: number;
	clientId: string | null;
	clientLabel: string;
	clientVersion: string | null;
	toolName: string;
	title: string | null;
	phase: McpActivityPhase;
	summary: string;
	startedAt: number;
	completedAt: number | null;
	durationMs: number | null;
	argumentsJson: string | null;
	resultJson: string | null;
	errorJson: string | null;
	requestId: string | null;
	sessionId: string | null;
	contentItems?: McpContentItem[];
};

export type McpActivityState = {
	sequence: number;
	eventCapacity: number;
	rows: McpActivityRow[];
	seenEventIds: string[];
};

function boundedTail<T>(items: T[], limit: number): T[] {
	return items.length <= limit ? items : items.slice(items.length - limit);
}

function correlationKey(event: McpActivityEvent): string {
	return event.request_id
		? `plugin:${event.request_id}:${event.tool_name}`
		: `plugin:event:${event.event_id}`;
}

function startedAt(event: McpActivityEvent): number {
	return event.phase === 'started'
		? event.timestamp_ms
		: Math.max(0, event.timestamp_ms - (event.duration_ms ?? 0));
}

function rowFromEvent(event: McpActivityEvent, current?: McpActivityRow): McpActivityRow {
	const terminal = event.phase !== 'started';
	return {
		id: current?.id ?? event.event_id,
		correlationKey: current?.correlationKey ?? correlationKey(event),
		correlationId: event.correlation_id ?? current?.correlationId ?? null,
		source: 'plugin',
		sequence: event.ingestion_sequence,
		clientId: event.client.id,
		clientLabel: event.client.label,
		clientVersion: event.client.version,
		toolName: event.tool_name,
		title: event.title,
		phase: event.phase,
		summary: event.summary,
		startedAt: current?.startedAt ?? startedAt(event),
		completedAt: terminal ? event.timestamp_ms : null,
		durationMs: terminal ? event.duration_ms : null,
		argumentsJson: event.arguments_json ?? current?.argumentsJson ?? null,
		resultJson: event.result_json ?? (terminal ? null : (current?.resultJson ?? null)),
		errorJson: event.error_json ?? (terminal ? null : (current?.errorJson ?? null)),
		requestId: event.request_id ?? current?.requestId ?? null,
		sessionId: event.session_id ?? current?.sessionId ?? null,
		contentItems: current?.contentItems
	};
}

function applyFreshEvent(state: McpActivityState, event: McpActivityEvent): McpActivityState {
	const key = correlationKey(event);
	const rows = state.rows.map((row) => ({ ...row }));
	const index = rows.findIndex((row) => row.source === 'plugin' && row.correlationKey === key);
	const current = index >= 0 ? rows[index] : undefined;
	const row = rowFromEvent(event, current);
	if (index >= 0) rows[index] = row;
	else rows.push(row);

	return {
		...state,
		sequence: event.ingestion_sequence,
		rows: boundedTail(rows, state.eventCapacity),
		seenEventIds: boundedTail([...state.seenEventIds, event.event_id], state.eventCapacity * 2)
	};
}

export function hydrateMcpActivity(snapshot: McpActivitySnapshot): McpActivityState {
	const eventCapacity = Math.max(
		1,
		snapshot.stream_health.event_capacity || snapshot.events.length || 1
	);
	let state: McpActivityState = {
		sequence: 0,
		eventCapacity,
		rows: [],
		seenEventIds: []
	};
	for (const event of snapshot.events) {
		if (event.ingestion_sequence <= state.sequence || state.seenEventIds.includes(event.event_id))
			continue;
		state = applyFreshEvent(state, event);
	}
	return { ...state, sequence: Math.max(state.sequence, snapshot.sequence) };
}

export function applyMcpActivityEvent(
	state: McpActivityState,
	event: McpActivityEvent
): McpActivityState {
	if (event.ingestion_sequence <= state.sequence || state.seenEventIds.includes(event.event_id)) {
		return state;
	}
	return applyFreshEvent(state, event);
}

export function mergeConsoleActivityRow(
	state: McpActivityState,
	row: McpActivityRow
): McpActivityState {
	const rows = state.rows.map((item) => ({ ...item }));
	const index = rows.findIndex((item) => item.source === 'console' && item.id === row.id);
	if (index >= 0) rows[index] = { ...row, source: 'console' };
	else rows.push({ ...row, source: 'console' });
	return { ...state, rows: boundedTail(rows, state.eventCapacity) };
}
